backend/src/routes/calibration.ts

```diff
@@ -0,0 +1,64 @@
+import {Connection, Repository} from 'typeorm'
+import {Request, RequestHandler, Response} from 'express'
+import {Calibration, CalibrationData} from '../entity/Calibration'
+
+
+export class CalibrationRoutes {
+
+  constructor(conn: Connection) {
+    this.calibRepo = conn.getRepository<Calibration>('calibration')
+  }
+
+  private calibRepo: Repository<Calibration>
+
+  calibration: RequestHandler = async (req: Request, res: Response, next) => {
+    this.findCalibration(req.query)
+      .then(result => {
+        if (result == undefined) return next({status: 404, errors: 'Calibration data not found'})
+        const calibData = result.calibration.sort((a, b) =>
+          new Date(b.createdAt).getTime() - new Date(a.createdAt).getTime())
+        if (req.query.showAll == undefined) {
+          res.send([calibData[0]])
+          return
+        }
+        res.send(calibData)
+      })
+      .catch(err => next({ status: 500, errors: err }))
+  }
+
+  postCalibration: RequestHandler = async (req: Request, res: Response, next) => {
+    const body = req.body
+    const calibData: CalibrationData = {
+      calibrationFactor: body.calibrationFactor,
+      createdAt: new Date()
+    }
+    try {
+      const existingCalib = await this.findCalibration(req.body)
+      if (existingCalib) {
+        existingCalib.calibration.push(calibData)
+        await this.calibRepo.update(existingCalib.id, existingCalib)
+        return res.sendStatus(200)
+      }
+      await this.calibRepo.insert({
+        site: body.site,
+        instrument: body.instrument,
+        measurementDate: body.date,
+        calibration: [calibData]
+      })
+      return res.sendStatus(200)
+    } catch (err) {
+      return next({ status: 500, errors: err })
+    }
+  }
+
+  private findCalibration(query: any) {
```

Comment:
Tämän pitäisi palauttaa aina viimeisin arvo. Ensimmäistä kalibrointia vanhemmista päivistä voi heittää virhettä.

Reply:
Tarkoitatko, että JOS taulusta ei löydy annetulla päivällä arvoa JA annettu päivä on suurempi kuin viimeisin taulussa oleva NIIN palauta taulun viimeisin arvo? Entäs tilanteessa, jossa taulusta ei löydy annetulla päivällä arvoa MUTTA taulusta löytyy aikaisemmalle ja myöhemmälle päivälle arvo? Palautetaanko silloin lähimmän päivämäärää edeltävän päivämäärän arvo? Ja sitten tilanteessa, jossa annettu päivämäärä on vanhempi kuin aikaisin taulusta löytyvä pvm, heitetään virhe.

Reply:
Ratkaisu tähän ongelmaan:
- Kun tehdään GET pvm:ään, joka on tuoreempi kuin yksikään kannasta löytyvä arvo, palautetaan lähin edellinen kannasta löytyvä kalibrointiarvo.
- Kun tehdään POST pvm:ään, joka on tuoreempi kuin yksikään kannasta löytyvä arvo, asetetaan kaikkien päivien välillä `[ edellinen asetettu arvo, POSTin pvm [` arvoksi lähin edellinen kannasta löytyvä kalibrointiarvo.

Näin ollen kaikilla päivillä tulee olemaan arvot, paitsi viimeisimmän päivityksen jälkeisillä päivillä. Täten yksittäisten menneisyydessä olevien arvojen päivitys ei johda odottamattomaan toiminnallisuuteen, vaan muokkaa vain kyseisen päivän arvoa.

Reply:
Kuulostaa järkevältä, tehdään näin.